src/symbolic_dimensions.jl

```diff
@@ -95,15 +106,122 @@ end
 expand_units(q::QuantityArray) = expand_units.(q)
 
 
-Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(data(d)))
-Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions) = data(l) == data(r)
-Base.iszero(d::SymbolicDimensions) = iszero(data(d))
-Base.:*(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) + data(r))
-Base.:/(l::SymbolicDimensions, r::SymbolicDimensions) = SymbolicDimensions(data(l) - data(r))
-Base.inv(d::SymbolicDimensions) = SymbolicDimensions(-data(d))
-Base.:^(l::SymbolicDimensions{R}, r::Integer) where {R} = SymbolicDimensions(data(l) * r)
-Base.:^(l::SymbolicDimensions{R}, r::Number) where {R} = SymbolicDimensions(data(l) * tryrationalize(R, r))
+Base.copy(d::SymbolicDimensions) = SymbolicDimensions(copy(getfield(d, :nzdims)), copy(getfield(d, :nzvals)))
+function Base.:(==)(l::SymbolicDimensions, r::SymbolicDimensions)
+    nzdims_l = getfield(l, :nzdims)
+    nzvals_l = getfield(l, :nzvals)
+    nzdims_r = getfield(r, :nzdims)
+    nzvals_r = getfield(r, :nzvals)
+    nl = length(nzdims_l)
+    nr = length(nzdims_r)
+    il = ir = 1
+    while il <= nl && ir <= nr
+        dim_l = nzdims_l[il]
+        dim_r = nzdims_r[ir]
+        if dim_l == dim_r
+            if nzvals_l[il] != nzvals_r[ir]
+                return false
+            end
+            il += 1
+            ir += 1
+        elseif dim_l < dim_r
+            if !iszero(nzvals_l[il])
+                return false
+            end
+            il += 1
+        else
+            if !iszero(nzvals_r[ir])
+                return false
+            end
+            ir += 1
+        end
+    end
+
+    while il <= nl
+        if !iszero(nzvals_l[il])
+            return false
+        end
+        il += 1
+    end
+
+    while ir <= nr
+        if !iszero(nzvals_r[ir])
+            return false
+        end
+        ir += 1
+    end
+
+    return true
+end
+Base.iszero(d::SymbolicDimensions) = iszero(getfield(d, :nzvals))
+
+# Defines `inv(::SymbolicDimensions)` and `^(::SymbolicDimensions, ::Number)`
+function map_dimensions(op::Function, d::SymbolicDimensions)
+    return SymbolicDimensions(copy(getfield(d, :nzdims)), map(op, getfield(d, :nzvals)))
+end
 
+# Defines `*(::SymbolicDimensions, ::SymbolicDimensions)` and `/(::SymbolicDimensions, ::SymbolicDimensions)`
+function map_dimensions(op::O, l::SymbolicDimensions{L}, r::SymbolicDimensions{R}) where {O<:Function,L,R}
+    zero_L = zero(L)
+    zero_R = zero(R)
+    T = typeof(op(zero(L), zero(R)))
```

Comment:
I'm a bit worried about type instabilities coming from this line if there is a failure by the compiler to propagate constants. e.g., for certain rational-like types they might have some complicated logic for `+` and the compiler could refuse to inline.

I don't see an immediate need for this, as `map_dimensions` is only used for `+` and `-` on `SymbolicDimensions` currently, so I propose to make this simply `promote_type(L, R)`. Any objections? `promote_type` is what is used for `map_dimensions` for `AbstractDimensions` anyways.

Reply:
(Just pushed this change... let me know if you feel strongly against it)

Reply:
`promote_type` won't give the correct result in general. For instance, `promote_type(Bool, Bool) === Bool` but `typeof(+(zero(Bool), zero(Bool))) === Int`. Therefore such computations are also used in eg Statistics: https://github.com/JuliaStats/Statistics.jl/blob/bfc6326267f6fb922707ac0465e44218ff475d46/src/Statistics.jl#L216

Reply:
IMO one should not be worried about the compiler here. It will also not cause any type instabilities, in the worst case there's one additional operation (but for common types the compiler should figure this out at compile time).

Reply:
Ah, damn, I see. Okay we can roll that back. Maybe we can have a function

```julia
@generated function promote_type_by_eval(f::F, ::Type{L}, ::Type{R}) where {F,L,R}
    T = F.instance(zero(L), zero(R))
    return :($T)
end
```

just in case the compiler doesn't feel like constant propagation for certain complicated rational types?

I guess we should also implement this for the general `map_dimensions`, if possible.

Reply:
No, generated functions should be just a last resort and there's absolutely no need for any concerns here.

Reply:
(Or just implementing it for `+` and `-` to be safe)

Reply:
I still don't understand why you would want to add or change anything. The code (with `op`) is completely fine, performant, and guaranteed to create arrays of the correct type.

Reply:
Ah sorry our messages crossed and I didn't see what you said. Yeah sounds good, I'll remove it.